### PR TITLE
PCM generation now uses the ROOT modulemap to build modules.

### PR DESCRIPTION
**Pull request by Teemperor**
This patch also untangles the generation code of PCMs and PCHs.

### CODE REVIEW

File: core/dictgen/src/rootcling_impl.cxx [hunk 1]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
+}
 
-      // Write the generated bitstream to "Out".
-      OS->write((char *)&Buffer.front(), Buffer.size());
+////////////////////////////////////////////////////////////////////////////////
+/// Returns true iff a given module (and its submodules) contains all headers
+/// needed by the given ModuleGenerator.
+/// The names of all header files that are needed by the ModuleGenerator but are
+/// not in the given module will be inserted into the MissingHeader variable.
+static bool moduleContainsHeaders(TModuleGenerator &modGen, clang::Module *module,
```
Comment (pcanal) [inline core/dictgen/src/rootcling_impl.cxx]: Coding convention calls for CamelCase for funcitons names.  thanks.

File: core/dictgen/src/rootcling_impl.cxx [hunk 2]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
-                   const std::string &fwdDeclnArgsToKeepString,
-                   const std::string &headersClassesMapString,
-                   const std::string &fwdDeclString,
-                   std::ostream &dictStream,
-                   bool inlineInputHeader)
+/// Write the AST of the given CompilerInstance to the given File while
+/// respecting the given isysroot.
+/// If module is not a null pointer, we only write the given module to the
+/// given file and not the whole AST.
+int writeAST(StringRef FileName, clang::CompilerInstance *CI, StringRef ISysRoot, clang::Module *module = nullptr)
```
Comment (pcanal) [inline core/dictgen/src/rootcling_impl.cxx]: please use CamelCase for function names. thanks.

File: core/dictgen/src/rootcling_impl.cxx [hunk 3]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
 
-      Writer.WriteAST(CI->getSema(), modGen.GetModuleFileName().c_str(),
-                      module, ISysRoot.c_str());
+   return 0;
+}
 
-      // Write the generated bitstream to "Out".
-      OS->write((char *)&Buffer.front(), Buffer.size());
+////////////////////////////////////////////////////////////////////////////////
+/// Returns true iff a given module (and its submodules) contains all headers
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: Typo `iff`

File: core/dictgen/src/rootcling_impl.cxx [hunk 4]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
+   }
+   return FoundAllHeaders;
+}
 
+////////////////////////////////////////////////////////////////////////////////
+/// Generates a module from the given ModuleGenerator and CompilerInstance.
+int GenerateModule(TModuleGenerator &modGen, clang::CompilerInstance *CI)
+{
+   // modGen must not be in PCH mode. Otherwise this is a programming error.
+   assert(!modGen.IsPCH());
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: Could you add `&& "Comment"` to the assert?

File: core/dictgen/src/rootcling_impl.cxx [hunk 5]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
+      auto LengthWithoutSuffix = ModuleName.size() - strlen("_rdict.pcm");
+      ModuleName = ModuleName.substr(0, LengthWithoutSuffix) + ".so";
+   }
+
+   // Actually lookup the module on the computed module name.
+   clang::Module *module = ModuleMap.findModule(ModuleName);
+
+   // Inform the user and abort if we can't find a module with a given name.
+   if (module == nullptr) {
+      std::cerr << "error: Couldn't find module with name '" << ModuleName << "' in modulemap!\n";
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: We should ROOT's error reporting facilities, eg `Error`.

File: core/dictgen/src/rootcling_impl.cxx [hunk 6]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
+      return 1;
    }
 
-   // Free up some memory, in case the process is kept alive.
-   Buffer.clear();
+   // Check if the loaded modules covers all headers that were specified
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: Typo: 'cover'.

File: core/dictgen/src/rootcling_impl.cxx [hunk 7]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
+   // by the user on the command line. This is an integrity check to
+   // ensure that our used module map is
+   std::vector<std::string> MissingHeaders;
+   if (!moduleContainsHeaders(modGen, module, MissingHeaders)) {
+      // FIXME: Upgrade this to an error once modules are stable.
+      std::cerr << "warning: Couldn't find the following specified headers in "
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: Likewise, use `Error`.

File: core/dictgen/src/rootcling_impl.cxx [hunk 8]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
+                << "the module " << module->Name << ":\n";
+      for (auto &H : MissingHeaders) {
+         std::cerr << "  " << H << "\n";
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: As above.

File: core/dictgen/src/rootcling_impl.cxx [hunk 9]
```diff
@@ -2187,85 +2188,148 @@ static bool InjectModuleUtilHeader(const char *argv0,
...
+         MissingHeaders.push_back(H);
+         FoundAllHeaders = false;
+      }
+   }
+   return FoundAllHeaders;
+}
 
+////////////////////////////////////////////////////////////////////////////////
+/// Generates a module from the given ModuleGenerator and CompilerInstance.
+int GenerateModule(TModuleGenerator &modGen, clang::CompilerInstance *CI)
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: What about returning 'bool', it seems we check only for success.

**Comment by vgvassilev** [comment]
@martinmine, it looks like the build failed due to configuration issues. Could you have a look, this is on a separate branch called `runtime-cxxmodules`.

File: core/dictgen/src/rootcling_impl.cxx (hunk 3 above)
Comment (Axel-Naumann) [inline core/dictgen/src/rootcling_impl.cxx]: https://www.wikiwand.com/en/If_and_only_if :-)

File: core/dictgen/src/rootcling_impl.cxx (hunk 5 above)
Comment (Axel-Naumann) [inline core/dictgen/src/rootcling_impl.cxx]: `ROOT::TMetaUtils::Error` to be more precise.

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Pull request merged**